Sources/AWSLambdaRuntimeCore/LambdaContext.swift

```diff
@@ -117,9 +127,9 @@ extension Lambda {
             logger[metadataKey: "awsRequestID"] = .string(requestID)
             logger[metadataKey: "awsTraceID"] = .string(traceID)
             var baggage = BaggageContext()
-            // TODO: handle error
+            // TODO: use `swift-tracing` API, note that, regardless, we can ONLY extract X-Ray Context
```

Comment:
not really; "we" don't decide what we extract, the configuration of instruments decides that, and yes since xray would be configured it'd extract it's own context here.

Specifically: 

`tracer.extract(<from where to extract, could be a dictionary> / http headers etc, into: &baggage, using: extractor apropriate to the first parameter, so http headers or similar)`

Reply:
the problem here is that only the X-Ray trace context is provided by Lambda Runtime API (in header), see https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html#runtimes-api-next

the context of other instruments may be provided in the payload BUT
the payload is not handled by the Lambda runtime but by (user provided) lambda event handler.

example:

the source event may be API Gateway which will contain HTTP headers.

However the `AWSLambdaRuntimeCore` requires only the event to be `Decodable` but the type itself is defined outside.
So lambda event handler (implemented by user) may parse the HTTP headers in the payload of API Gateway event
and extract context from them, but this will have to happen outside (done by user).

Reply:
so to make it sorter:

`XRayInstrument` can use the `Instrument.extract` API (its implemented and tested)
but lambda runtime has only information about the X-Ray trace context.

Other context can be provided by payload and needs to be handled by user)

Reply:
I see, I missed that bit of the Lambda design. So in this integration style, even if one triggered the lambda via an http request, it would not get “the http request” but just the body, and the headers are the ones as listed on there, including the XRay trace header etc.

There AFAIR exists an integration mode though to get the entire request, right?

> Pass through the entire request – A Lambda function can receive the entire HTTP request (instead of just the request body) and set the HTTP response (instead of just the response body) using the AWS_PROXY integration type.

Is this something that the runtime currently is able to handle? Seems more to be about how the API Gateway is configured right? Though I’ve not had the time to dig deeper into this yet.

You’d probably know more about this @fabianfett, we can catch up about this today maybe?

Reply:
I guess you are referring to [API Gateway Lambda proxy integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html)

This does affect Lambda custom runtime API, it affects API Gateway v1 "REST API" routing which, if configured that way, does not try to route events based on a RESTful model, instead it forwards all events to lambda which needs to resolve HTTP method, path and arguments itself:

> ANY /{proxy+}: The client must choose a particular HTTP method, must set a particular resource path hierarchy, and can set any headers, query string parameters, and applicable payload to pass the data as input to the integrated Lambda function.

Note that:
- events with HTTP requests created by API Gateway may have different syntax, specifically:
  - API Gateway v2 "HTTP API", see [AWSLambdaEvents/APIGateway+V2](https://github.com/swift-server/swift-aws-lambda-runtime/blob/main/Sources/AWSLambdaEvents/APIGateway%2BV2.swift)
  - API Gateway v1 "REST API", see [AWSLambdaEvents/APIGateway](https://github.com/swift-server/swift-aws-lambda-runtime/blob/main/Sources/AWSLambdaEvents/APIGateway.swift)
- event types are not defined in `AWSLambdaCore` but in `AWSLambdaEvents ` which, unlike `AWSLambdaCore`, does have dependency on `Foundation` at the moment.
- AWS does not limit context propagation to API Gateway (but it will not propagate context for other instruments unless it copies headers from the original requests and includes them in event payload, which is true for API Gateway)

Reply:
Thanks for explaining this in more depth @pokryfka, I need to read up some more here about aws/lambda in general it seems.